File: main/python/algorithms/heap_sort.py
def heapify(arr, n, i):
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2
    if l < n and arr[largest] < arr[l]:
        largest = l
    if r < n and arr[largest] < arr[r]:
        largest = r

    if largest != i:
        temp = arr[i]
        arr[i] = arr[largest]
        arr[largest] = temp
        heapify(arr, n, largest) 
 
def heap_sort(arr):
    n = len(arr)
 
    for i in range(n//2 - 1, -1, -1):
        heapify(arr, n, i)
 
    for i in range(n-1, 0, -1):
        temp = arr[i]
        arr[i] = arr[0]
        arr[0] = temp
        heapify(arr, i, 0)

File: main/c++/algorithms/heap_sort.cpp
#include <cmath>

#include "sort.h"

void heapify(int *array, int len, int index) {
    int largest = index;
    int l = 2 * index + 1;
    int r = 2 * index + 2;
    if (l < len && array[largest] < array[l]) {
        largest = l;
    }
    if (r < len and array[largest] < array[r]) {
        largest = r;
    }
    if (largest != index) {
        int temp = array[index];
        array[index] = array[largest];
        array[largest] = temp;
        heapify(array, len, largest);
    }
}
 
void heap_sort(int *array, int len) {
 
    for (int i = floor(len/2) - 1; i > -1; i--) {
        heapify(array, len, i);
    }
 
    for (int i = len-1; i > 0; i--) {
        int temp = array[i];
        array[i] = array[0];
        array[0] = temp;
        heapify(array, i, 0);
    }
}